### ISSUE TITLE
php-fpm crashing every few day 

### ISSUE THREAD

**Issue by ghost** [perusio/drupal-with-nginx#55]
Good Morning,

I understand this is probably not related to this config, however you seem to be a wealthy source of information on nginx / php-fpm and drupal. Perhaps you have seen this problem before...

Every few days at exactly the same times 08:04, 20:04 (approx.) The php-fpm children of all of my web servers increase dramatically until the max_children is reached, the servers then come under heavy load and syslog show kernel messages related to oomkiller. php-fpm then appears to crash/restart all processes and the site then loads fine again for another few days.

This cycle repeats every few days at the same time, it also occurs whether I set max_children to 10, 50, 100 etc. what ever the value, they spawn more until php-fpm crashes, some kind of memory leak or infinite loop?

Site traffic is also minimal at these times and the site can handle 200% more traffic at other peak times without problems.

Versions:
Ubuntu 11.10 - 3.0.0-12-server x86_64
nginx 1.0.5-1
php 5.3 5.3.6-13ubuntu3.3
drupal6

Any ideas would be well appreciated.

Regards,

Alun.

**Comment by tdm4** [comment]
Using fastcgi_keep_conn on; in Nginx and using PHP-FPM in 'ondemand' or 'dynamic' configuration is very bad. Once php-fpm terminates the php-fpm child, nginx still thinks it is connected to it and after a while php-fpm will just keep forking new children, hit the limit, and fall over.

I have seen this on Drupal sites and non-Drupal sites (like Magento). In the drupal config I noticed this directive is on. This can pose major problems. It took make a long long time to figure out that this setting was a problem. I recommend having it turned off in the config, always. The supposed performance benefit isn't worth the grief of wondering why 20-30 php-fpm children died with SIGTERM 15.. and googling for the answer gives hardly any answers. I had to track the problem down to here:

http://mailman.nginx.org/pipermail/nginx/2013-February/037496.html
https://bugs.php.net/bug.php?id=60961
https://bugs.php.net/bug.php?id=63395

**Comment by cthshabel** [comment]
tdm4, you are the only person that has really explained this the right way for me.

I have always struggled with using "fastcgi_keep_conn on;" in the perusio config. I'm surprised more people don't complain of the errors.  Once I comment it out, everything works fine.

btw i am using 2 nginx instances (1 reverse proxy for ssl termination in front and 1 web server for drupal environment in back).

today, i tried to use the drupal.conf and tried using the microcache proxy (sounded like an awesome idea!), and it always returns "nginx: [emerg] the shared memory zone "microcache" is already declared for a different use in /etc/nginx/apps/drupal/microcache_proxy_auth.conf:7".

will investigate a little more!

**Comment by perusio** [comment]
@tdm4 I agree that for **on demand** the keepalive setup can pose problems, but not for dynamic. Per the links you posted to the php-fpm bug reports, only and only if you have a number of connections in the connection pool **greater than** the minimum number of servers. Of course if you have **4** as the minimum number of servers and you have `keepalive 5` in the upstream definition this may pose problems. I will add something to the README and also comments on the files to warn about that.

So bottom line: 

* **dynamic**: in the upstream set the number of connections in the connection pool (for keepalive) to a number **less ir equal** to the minimum number of servers in php-fpm.

* **on demand**: disable the connection pool. Comment out the `keepalive 5;` line in the upstream definition and comment out the `fastcgi_keep_conn on;` line.

**Comment by perusio** [comment]
@cthshabel You cannot declare memory zone twice. your including the cache zone definition twice. Comment out the inclusion of the file in the **second vhost** I'm sure you mean vhost and not **instance**. A second instance means a different master process and another PID for it. If it was really a second instance it wouldn't complain. What do you mean by _second instance_?

**Comment by tdm4** [comment]
@perusio Re: Your comments above. With dynamic, the keepalive must be set to a number less than or equal to the minimum servers in php-fpm. Is that pm.start_servers or pm.min_spare_servers? Also, fastcg_keep_conn should be set to on with php-fpm dynamic but not for ondemand?

What happens when one of the minimum servers nginx is expecting to talk to decides to close (presumably after pm.max_requests and a new one spins up?)

We've had keepalive not set, but fastcgi_keep_conn on; and it basically blows up php-fpm after a certain amount of time. So perhaps having keepalive not set and fastcgi_keep_conn on; messes things up as well?

**Comment by cjbirk** [comment]
Hello.  I just ran into this issue recently with a client.  It appears that they installed (but did not properly configure) Apache Solr - some sort of search platform.

I figured that I would put this here just in case anyone else runs into this issue.  Basically this boils down to PHP error logging being extremely terrible. I hope someone finds this useful.

**Comment by pprishchepa** [comment]
@paradoxni try to disable `pm.max_requests` in PHP-FPM config (if enabled).
